src/main/java/com/google/devtools/build/lib/syntax/EvalUtils.java

```diff
@@ -74,6 +74,11 @@ private int compareLists(SkylarkList o1, SkylarkList o2) {
         @Override
         @SuppressWarnings("unchecked")
         public int compare(Object o1, Object o2) {
+          // optimize the most common case
```

Comment:
Is it really the most common case? In my experience integers are very infrequently used in Starlark programs (because list iteration doesn't use explicit indices) and string comparisons are much more important.

Reply:
I think so, for example, this is a grep in Bazel:

```
% rg 'if.*(>|<)=? \d' -t bazel
src/rule_size_test.bzl
57:    if ctx.attr.expect < 0:
60:    if ctx.attr.margin < 0 or ctx.attr.margin > 100:
71:    if ctx.attr.margin > 0:

tools/osx/xcode_configure.bzl
34:    if (prefix_index < 0):
38:    if (suffix_index < 0):

tools/cpp/cc_toolchain_config_lib.bzl
261:    if len(flags) > 0 and len(flag_groups) > 0:

tools/cpp/windows_cc_configure.bzl
336:    if len(versions) < 1:

tools/android/android_sdk_repository_template.bzl
88:        if api_level >= 23:
96:        if api_level >= 28:

third_party/grpc/build_defs.bzl
21:    if len(ctx.attr.srcs) > 1:

tools/build_defs/repo/utils.bzl
91:    if len(ctx.attr.patches) > 0 or len(ctx.attr.patch_cmds) > 0:
209:                if len(w.strip()) > 0
274:        if len(schemerest) < 2:

tools/build_defs/pkg/pkg.bzl
107:        if dotPos > 0:

tools/build_defs/pkg/path.bzl
44:        if len(data_path) >= 2 and data_path[0:2] == "./":

third_party/protobuf/3.6.1/protobuf.bzl
427:    if current.compare_to(expected) < 0:

src/main/res/win_res.bzl
21:    if i > 0:
```

And I couldn't find string any comparison in Bazel Bazel definitions.

OK, maybe strings are special too, but string comparison is much more expensive than an integer comparison, so the special case for strings won't be that helpful.

(I must admit, I'm not sure how practically important is this optimization)

Reply:
I meant: is Integer the most common case dynamically. String comparisons tend not to be explicit; most come from functions like `sorted`. In the Go implementation of a workload equivalent to Bazel's loading phase, string comparisons and equality tests were far more numerous than those for integers.

Reply:
Interesting.

Similar optimization for strings:

```
          if (o1 instanceof String && o2 instanceof String) {
            return ((String) o1).compareTo((String) o2);
          }
```

makes this benchmark:

```
def foobar():
    return sorted(["ab", "cde", "zzb", "aab", "aaa2", "zzz2", "ffv2"])

def bench():
    for i in range(1000000):
        foobar()

bench()
```

4 (!) times faster.

Reply:
Great, let's have both. Put string first, above all else.

Reply:
Done!